sstables_loader.cc

```diff
@@ -192,7 +195,18 @@ class tablet_sstable_streamer : public sstable_streamer {
 };
 
 inet_address_vector_replica_set sstable_streamer::get_endpoints(const dht::token& token) const {
-    return get_all_endpoints(token);
+    return get_all_endpoints(token) | std::views::filter([&topo = _erm->get_topology(), scope = _stream_scope] (const auto& ep) {
+        switch (scope) {
+        case stream_scope::all:
+            return true;
+        case stream_scope::dc:
+            return topo.get_datacenter(ep) == topo.get_datacenter();
+        case stream_scope::rack:
+            return topo.get_rack(ep) == topo.get_rack();
```

Comment:
I think it's dangerous to reduce the scope during migration since a tablet replica may be migrated across rack (when adding a rack), and of course across nodes. If a tablet is migrated from node A to B, we might miss the streaming to pending replica in another DC or rack, and once migration completes, the data is missing since it's cleaned up on leaving. Maybe we need to wait for topology to be quiesced like we did for repair, when reducing the scope. We already take the erm, but it's not enough since a migration can be running already. Currently load and stream works with migrations since it writes to both leaving and pending replicas.

Reply:
It's not migration code, it's load-and-stream which is -- "nodetool refresh" and "restore from object storage"

Reply:
I know that, but what I was saying is that migration can run in parallel to load-and-stream, and once that happens, you have that problem.

Reply:
see https://github.com/scylladb/scylladb/commit/771cbf9b79d10b43217371874719b0fc57cc1228

Reply:
load-and-stream received tablet treatment to work in parallel with migrations. holding e__r__m prevents migration from advancing, but doesn't prevent l-a-s from running while a migration is going on.

Reply:
We should probably stream to both source and destination during migration.
There's also the rebuild transition where we should also stream to the pending replica.

Reply:
> We should probably stream to both source and destination during migration. There's also the rebuild transition where we should also stream to the pending replica.

exactly. that's what we do today, so we should just preserve it. we should be safe by the default, so even if we restrict scope to current node, yet migration is happening from this node to another, we still favor correctness.

Reply:
local node scope is meaningless without freezing migrations.

Rack is okay, iff RF=#racks (otherwise migrations can change racks)
dc is okay, iff NetworkTopologyStrategy
primary-replica-only, iff vnodes and no topology changes
all is okay, but wasteful

Reply:
I don't understand why we want node scope. It sounds to me like a step back, restoring old refresh behavior which load and stream intended to fix. With the dynamics of tablets it's very dangerous to reduce destination scope to node or even rack since tablets can escape the attempt to stream to them. For example, when you run l-a-s with node scope at node A, a tablet lives in node B, and when you run it at node B, that tablet migrates back to A. So it escaped your attempt to stream to it. Even if you freeze migrations in each run, that's not enough since migrations can happen between runs. Even rack scope is dangerous since tablets can be migrated between racks. It sounds to me the safe options are the existing primary replica option and the new Dc scope since balancer won't migrate across dcs. Why can't we work with them? What's the motivation for node and rack scopes?

Reply:
>> I don't understand why we want node scope.
> I think it can be useful for 1-1 restore when the cluster is restored to an equivalent topology  ...

Yup, that's the goal. Maybe propagating this "scope" to nodetool refresh is not good idea indeed, but it was supposed for restore, and in this case migration can (and should) be disabled

Reply:
> With file-based streaming this can be optimized (in a safe way) to just copy/move the sstable files to the base directory, but will still enjoy the safety protections of load-and-stream (incl token range checks for tablets, and other sanity checks like digest verification, etc.)

This is what `stream_fully_contained_sstables()` is supposed to take care of

Reply:
Rethinking the `node` scope.  We also have the original `nodetool refresh` api, without --load-and-stream, that loads sstables only locally.
Currently it filters the sstable partition using "cleanup" and resharding, so to leave only the partitions that are owned by the local node segregated into sstables by shard ownership.
I think that's equivalent to the `node` scope.

Reply:
We should be fine correctness wise with whichever scope as long as stream to pending replica (already existing behavior with both "all" and "primary-replica"), like we discussed initially. We might redundantly load data to same tablet replica if we don't disable migrations throughout the restore, but at least we don't miss any tablet. 
I don't think we should assume user disabled migration, it's better to be correct if the user forgot to do so, by streaming to pending replica (as we do today). If migration was disabled, then we lose nothing. If it wasn't, we miss data.

Perhaps once we switch 100% to tablets, file streaming will be sufficient to satisfy throughput requirements on 1-1 restore (we just need to make sure that table is created with same tablet count as in the backup), without pushing the responsibility of optimization to something external. Playing games always bring some risk.

Reply:
@xemul :point_up: 
This is not resolved.
Can do `topo.get_location(ep) == topo.get_location()`